### PR TITLE
Make copying data optional

### PR DESCRIPTION
**Pull request by erikprice**
The context for this PR is in #424 .

### CODE REVIEW

File: SocketRocket/SRWebSocket.m [hunk 1]
```diff
@@ -997,7 +1014,9 @@ - (void)_readFrameContinue;
 - (void)_readFrameNew;
 {
     dispatch_async(_workQueue, ^{
-        [_currentFrameData setLength:0];
+        // Don't reset the length, since Apple doesn't guarantee that this will free the memory (and in tests on
+        // some platforms, it doesn't seem to, effectively causing a leak the size of the biggest frame so far).
+        _currentFrameData = [[NSMutableData alloc] init];
```
Comment (nlutsenko) [inline SocketRocket/SRWebSocket.m]: This by itself is worth a ton!
Indeed this is the only valid way to reset the data...

File: SocketRocket/SRWebSocket.h [hunk 2]
```diff
@@ -264,6 +264,18 @@ NS_DESIGNATED_INITIALIZER;
...
+ Send binary data to the server, without making a defensive copy of it first.
+
+ @param data  Data to send.
+ @param error On input, a pointer to variable for an `NSError` object.
+ If an error occurs, this pointer is set to an `NSError` object containing information about the error.
+ You may specify `nil` to ignore the error information.
+
+ @return `YES` if the string was scheduled to send, otherwise - `NO`.
+ */
+- (BOOL)sendWithoutCopyingData:(nullable NSData *)data error:(NSError **)error NS_SWIFT_NAME(sendWithoutCopying(data:));
```
Comment (nlutsenko) [inline SocketRocket/SRWebSocket.h]: Let's name it Swiftier and closer to Foundation in Swift 3.0 as well.
```
- (BOOL)sendDataNoCopy:(nullable NSData *)data error:(NSError **)error NS_SWIFT_NAME(send(dataNoCopy:));
```

**Comment by nlutsenko** [comment]
Let's talk about the `delegate` method for copying data.

After looking deeper into the calls to `handleFrameWithData:opCode:` - there are numerous cases, where reading from that data would be incredibly unsafe, since we mutate the data and `NSData` is not thread-safe.I can see the value from not copying the incoming data, but at the same time using un thread-safe data can lead to crazy crazy problems on the receiving end and at the current moment.

Let's start with just the sending end of it and make this PR just about that, so we are unblocked on this PR.

For non-copying the frame data - looks like we can actually not copy the frame data, unless it's a web socket control frame, since we are going to simply re-create the `currentFrameData` again, so I bet there is a flow that we can figure out to be smart about copying, without requiring to implement the delegate method or configuration change.

**Comment by nlutsenko** [comment]
Awe. Are you going to send a new one?

**Comment by erikprice** [comment]
Yup, that PR was targeting your old `no-errors` branch. New one going to target `master`.

**Pull request closed without merging**